tests/xdc/history_replication_dlq_test.go

```diff
@@ -0,0 +1,454 @@
+func (s *historyReplicationDLQSuite) TestWorkflowReplicationTaskFailure() {
+	// This test uses channels to synchronize between the main test goroutine and the background task processing for
+	// replication, so we use a context with a timeout to ensure that the test doesn't hang forever when we try to
+	// receive from a channel.
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, testTimeout)
+	defer cancel()
+
+	// Register a namespace.
+	ns := "dlq-replication-test-namespace"
+	_, err := s.cluster1.GetFrontendClient().RegisterNamespace(ctx, &workflowservice.RegisterNamespaceRequest{
+		Namespace:                        ns,
+		Clusters:                         s.clusterReplicationConfig(),
+		ActiveClusterName:                s.clusterNames[0],
+		IsGlobalNamespace:                true,
+		WorkflowExecutionRetentionPeriod: timestamp.DurationPtr(1 * time.Hour * 24), // required param
+	})
+	s.NoError(err)
+
+	// Create a worker and register a workflow.
+	activeClient, err := sdkclient.Dial(sdkclient.Options{
+		HostPort:  s.cluster1.GetHost().FrontendGRPCAddress(),
+		Namespace: ns,
+	})
+	s.NoError(err)
+	tq := "dlq-replication-test-task-queue"
+	worker := sdkworker.New(activeClient, tq, sdkworker.Options{})
+	myWorkflow := func(ctx workflow.Context) (string, error) {
+		return "hello", nil
+	}
+	worker.RegisterWorkflow(myWorkflow)
+	s.NoError(worker.Start())
+	defer worker.Stop()
+
+	// Wait for the namespace to be replicated.
+	for {
+		var task *replicationspb.NamespaceTaskAttributes
+		select {
+		case task = <-s.namespaceReplicationTaskExecutorParams.tasks:
+		case <-ctx.Done():
+			s.FailNow("timed out waiting for namespace replication task to be processed")
+		}
+		if task.Info.Name == ns {
+			break
+		}
+	}
+
+	// Execute the workflow.
+	run, err := activeClient.ExecuteWorkflow(ctx, sdkclient.StartWorkflowOptions{
+		TaskQueue: tq,
+		ID:        workflowIDToFail,
+	}, myWorkflow)
+	s.NoError(err)
+
+	// Wait for the workflow to complete.
+	var result string
+	err = run.Get(ctx, &result)
+	s.NoError(err)
+	s.Equal("hello", result)
+
+	// Wait for the replication task executor to process the replication task.
+	select {
+	case <-s.replicationTaskExecutorParams.tasks:
+	case <-ctx.Done():
+		s.FailNow("timed out waiting for replication task to be processed")
+	}
+
+	// Wait for at least one replication task to be added to the DLQ. There could be more, but we only care about the
+	// first one because they should all be for this workflow.
+	var request *persistence.PutReplicationTaskToDLQRequest
+	select {
+	case request = <-s.executionManagerParams.dlqRequests:
+	case <-ctx.Done():
+		s.FailNow("timed out waiting for replication task to be added to DLQ")
+	}
+	s.Equal(s.clusterNames[0], request.SourceClusterName)
+	s.Equal(1, int(request.ShardID))
+
+	// Create a temporary file for TDBG output
+	file, err := os.CreateTemp("", "*")
+	s.NoError(err)
+	s.T().Cleanup(func() {
+		s.NoError(os.Remove(file.Name()))
+	})
```

Comment:
You can use https://github.com/temporalio/temporal/blob/main/tests/testutils/io.go#L36 here, I think

Reply:
Thanks for the tip! Using that now.